### PR TITLE
fix(pm): glob sprint plans into the Confluence sync, de-duplicate the file list [KAN-187]

### PR DESCRIPTION
**Pull request by adamtasteslikegood**
## What broke

`sync_pm_documents` and the watchdog observer iterated a hardcoded 7-path list in `pm_daemon.py`. It named `specs/SPRINT_0_PLAN.md` — a **182-byte TODO stub** — and omitted `SPRINT_1..4_PLAN.md`, ~91 KB of real sprint planning.

So the placeholder synced to Confluence for four sprints, and every actual sprint plan was skipped.

| File | Size | Was syncing |
|---|---|---|
| `SPRINT_0_PLAN.md` | 182 B (empty stub) | ✅ yes |
| `SPRINT_1_PLAN.md` | 17 KB | ❌ no |
| `SPRINT_2_PLAN.md` | 16 KB | ❌ no |
| `SPRINT_3_PLAN.md` | 27 KB | ❌ no |
| `SPRINT_4_PLAN.md` | 31 KB | ❌ no |

Silently, which is the worse half: `sync_pm_documents` reports what it synced and never what it skipped, so it returned "Sync successful" every time. That is how PR #3313's amended Sprint 4 plan was merged, reported as synced, and never reached Confluence.

## Secondary defect: the list was duplicated 4x and had already drifted

1. `scripts/pm/pm_daemon.py` — 7 entries (de-facto source of truth)
2. `.claude/hooks/sessionstart-pm-briefing.sh` — inline copy in a separate Python heredoc
3. `scripts/pm/atlassian_pm_link.py` — a **different** 4-entry subset (already drifted)
4. Docs: `CLAUDE.md`, `docs/PM_TOOLING.md`, `.claude/skills/run-pm-daemon/SKILL.md`

`docs/PM_TOOLING.md` additionally documented the tool as syncing *"all `specs/*.md` files"* — false, and precisely why the omission read as surprising rather than expected.

This is the failure mode `AGENTS.md` was collapsed to prevent.

## Fix

- **New `scripts/pm/_canonical_pm_files.py`** — the single source of truth. Curated list **plus a `specs/SPRINT_*_PLAN.md` glob**, so future sprints need no code change. Enumerating `SPRINT_4_PLAN.md` would have fixed today and re-broken at Sprint 5.
- **Stdlib-only by contract.** The briefing hook imports it under the system `python3`, outside the daemon's venv. A test asserts this via AST — a third-party import here would make the fail-open hook silently stop briefing.
- **Resolution moved to call time** against a root. The daemon is long-lived; a sprint plan created mid-session is now picked up without a restart, where the cached `__init__` list needed one.
- **The hook imports that module** instead of keeping its own copy, with a fail-open fallback so a checkout without `scripts/` still briefs rather than going silent.
- **`sync_pm_documents` now names the denominator** ("Considered 11 canonical PM file(s)"), so an omission is visible instead of indistinguishable from success.
- **Docs corrected** — the false "all `specs/*.md`" claim plus the three stale lists.

## Verification

Not just "tests pass" — proven to fail for the right reason:

- **71 tests green** (was 59; +12 new), `python3 -m unittest discover -s scripts/pm`.
- **Negative test:** neutering `SPRINT_GLOBS` to `[]` produces **6 failures**, including both end-to-end tests that actually spawn the hook. Comparing the two lists textually would have passed even with the hook unable to import the module at all — so the drift guard runs the real hook and diffs its output against the module.
- **Live run against production Confluence:** `sync_pm_documents` created `Sprint 1 Plan` (49414146), `Sprint 2 Plan` (49610753), `Sprint 3 Plan` (49381398), `Sprint 4 Plan` (49479691). Verified by direct API lookup, not the CQL search index — the index lagged several seconds behind the writes and initially showed only 2 of 4.

## Scope

Sprint plans only. `ux-backlog.md`, `KAN-119_LOOP_PLAN.md`, `CANONICAL_RECIPES_ROLLOUT.md` deliberately stay out (Adam's call, 2026-07-30).

## Lane

Labelled `not-sprint-4`. Sprint 4 is single-lane on KAN-155; this is out-of-lane tooling work Adam directed explicitly, recorded rather than silently widening the sprint the R1 risk exists to protect.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01RJTYg2F4r5Bcy6aX6hW8Ga

---
Rovo Dev code review: <strong>Rovo Dev has reviewed this pull request</strong>
Any suggestions or improvements have been posted as pull request comments.

### CODE REVIEW

**Review by adamtasteslikegood — COMMENTED** [review]
## Overview

Consolidates the PM-file list that was hardcoded in `pm_daemon.py` and copied inline into the SessionStart hook into a single stdlib-only module, `scripts/pm/_canonical_pm_files.py`. Sprint plans are now matched by glob instead of enumerated, resolution happens at call time (not import time), and `sync_pm_documents` now reports the size of the considered set so an omission is visible. Backed by a genuinely thorough test suite (unit tests, a negative test that neuters the glob, an AST check enforcing stdlib-only imports, and an end-to-end test that spawns the real hook rather than diffing two lists textually).

## Strengths

- `canonical_pm_files()`/`page_title_for()` are clean, well-documented, and the module docstring explains *why* each design decision (glob not enumeration, call-time not import-time resolution, stdlib-only) rather than just what it does.
- `test_hook_and_module_resolve_the_same_set` actually runs the bash hook and greps its output — this is the right way to guard against hook/module drift; a textual list comparison would pass even if the hook's import silently failed.
- Fail-open fallback in the hook (`canonical_pm_files is None` branch) preserves the "never go silent" property if `scripts/pm/_canonical_pm_files.py` isn't found.
- Numeric sprint sort (`SPRINT_10` after `SPRINT_2`, not between `SPRINT_1`/`SPRINT_2`) and `SPRINT_0` dedup between curated+glob are both correct and covered.

## Issues

**1. `scripts/pm/atlassian_pm_link.py` still has its own drifted list — this is one of the "4x" duplicates the PR describes as fixed.**

The PR body's "Secondary defect" section names four duplicated copies of the file list: `pm_daemon.py`, the SessionStart hook, `atlassian_pm_link.py` ("a **different** 4-entry subset — already drifted"), and docs. This diff fixes the first two and the docs, but `atlassian_pm_link.py:41-46` is untouched:

```python
LOCAL_PM_FILES = [
    "specs/planning_notes.md",
    "specs/plan.md",
    "specs/roadmap.md",
    "specs/design-plan.md",
]
```

still missing `SCRUM_BOOTSTRAP_AND_BOARD_PLAN.md`, `ATLASSIAN_PM_LINK.md`, and every sprint plan, and still not importing `_canonical_pm_files`. It feeds `summarize_local_pm_files()`, consumed by `sync_jira_confluence_status.py` (`pm:status`) — lower stakes than the Confluence-sync bug since it's read-only, but it means a third, independently-wrong copy of "the canonical set" survives this PR, and the PR description reads as if all three code copies were unified. Worth either migrating this file too or explicitly calling out in the PR body that it's deliberately out of scope (the way `ux-backlog.md` etc. are).

**2. The new test suite isn't wired into CI.**

`python3 -m unittest discover -s scripts/pm` isn't invoked anywhere in `.github/workflows/*.yml` (checked all of them — no reference to `scripts/pm`, `unittest`, or `pm_daemon`). The PR leans heavily on "71 tests green" and a negative test as proof of correctness, which is real evidence for *this* diff, but nothing stops the next edit to `pm_daemon.py` or the hook from silently re-breaking `canonical_pm_files()` — which is exactly the failure mode (silent, no CI signal) that produced KAN-187 in the first place. Since `pr-gate.yml` already has a Python job (`backend-test`, via `uv`), adding a lightweight `python3 -m unittest discover -s scripts/pm` step (no venv needed per the module's own stdlib-only contract) seems like a small, high-leverage addition — possibly worth a fast-follow if out of scope here.

## Minor

- `PMFileEventHandler._is_canonical_pm_file` now recomputes `canonical_pm_files(self.workspace_dir)` (a `specs/SPRINT_*_PLAN.md` glob + stats on the curated list) on every `on_modified` event, and the observer is scheduled with `recursive=True` over the entire `workspace_dir` (not just `specs/`), so every file save anywhere in the repo tree triggers a fresh directory glob rather than a check against the previously-cached list. Almost certainly negligible in practice (one small directory, not a recursive walk), but it is a real trade against the old cached-list version, made in exchange for the correct call-time-resolution property — just flagging it as a conscious trade-off rather than a free one.

## Test coverage / correctness

No correctness issues found in `canonical_pm_files`, `_sprint_sort_key`, or `page_title_for` — dedup, numeric sort, and existence-filtering all check out against the test suite's own scenarios plus a couple of edge cases I traced by hand (non-numeric sprint suffix, `SPRINT_0` curated+glob overlap).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by adamtasteslikegood** [comment]
## All three findings verified and fixed in `d2f95ab`

Two of them were introduced by this PR, which is worth stating plainly rather than filing as polish.

### 1. `LOCAL_PM_FILES` — the duplicate this PR's own body named

Correct, and the sharpest of the three. `scripts/pm/atlassian_pm_link.py:41` still held the drifted 4-entry list I listed as duplicate #3 in the description and then didn't migrate. Your trace of the consequence is right and is the part that makes it more than cosmetic: `summarize_local_pm_files()` feeds `PROJECT_PM_BRIEFING.md`, and the SessionStart hook **prefers** that file over the raw scan — so on any session after a briefing refresh, the briefing came from the drifted copy while the sync path covered everything.

Took your option 1, with the refinement you guessed at: the narrowness is *correct* — the briefing runs under a 12k-char cap and the sprint plans alone are ~91 KB, so including them would evict everything else. What was wrong was that it was a second **definition**. Promoted to `BRIEFING_SUMMARY_FILES` in `_canonical_pm_files.py`, so both sets live in one file. Two different sets is fine; two definitions is not. Verified `summarize_local_pm_files()` returns the identical four files as before.

### 2. Watcher perf — a regression this PR introduced

Confirmed: `observer.schedule(..., recursive=True)` at `pm_daemon.py:603`, so `on_modified` fires for every file event in the repo, and moving resolution to call time put a glob + N stats on that path. Not a pre-existing cost — I added it.

Added `could_be_canonical()`, a no-I/O basename reject that runs before anything touches disk. One deliberate departure from your snippet: I derived it from `CURATED_PM_FILES`/`SPRINT_GLOBS` via `fnmatch` rather than a hand-written `r"^SPRINT_\d+_PLAN\.md$"` in `pm_daemon.py`. A regex there would be a *new* copy of the exact pattern this module exists to hold once — it would have fixed the perf bug by re-committing the bug the PR is about.

Mid-session pickup is preserved. The test that matters is `test_never_rejects_what_canonical_pm_files_accepts`: a false negative in the fast path would silently stop syncing a real plan, which is worse than the perf issue it fixes.

### 3. Fallback sort mismatch

Fixed rather than documented. It's deep-tail as you say, but a *silently reordered briefing* is precisely the class of difference nobody thinks to check — and the module already has a test asserting numeric order, so leaving the fallback lex-sorted meant two documented-as-equivalent paths that aren't. Same numeric key now, pinned by a test that forces the fallback branch by running the hook against a checkout with no `scripts/pm`.

---

**Verification:** 8 new tests (71 → 79). Mutation-checked — dropping sprint plans from the fast path and removing the fallback sort key fails 4 tests. Hook re-verified end-to-end (11 files), `atlassian_pm_link` imports and summarises unchanged.

On the advisory review's summary: agreed on the ranking, and 1 and 2 were correctly identified as needing author judgment rather than a pushed fix — but both turned out to be defects rather than trade-offs once traced, so both are now fixed rather than accepted.

_Replied by Claude on Adam's behalf_

File: scripts/pm/pm_daemon.py [hunk 1]
```diff
@@ -125,15 +100,25 @@ def _find_confluence_page_id(title: str) -> str | None:
...
             return False
-        return any(resolved == candidate.resolve() for candidate in self.canonical_files if candidate.exists())
+        for relative_path in canonical_pm_files(self.workspace_dir):
+            candidate = self.workspace_dir / relative_path
+            try:
+                if resolved == candidate.resolve():
+                    return True
+            except FileNotFoundError:
+                continue
+        return False
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply scripts/pm/pm_daemon.py:129]: Fixed in `d2f95ab`, and the diagnosis was right about the mechanism: `observer.schedule(..., recursive=True)` at `pm_daemon.py:603` means `on_modified` fires for every file event in the repo, and moving resolution to call time put a glob + N stats on that path. Worth being plain that this was a regression **this PR introduced**, not a pre-existing cost — the pre-PR version compared against a list precomputed in `__init__`.

Added `could_be_canonical()`, a no-I/O basename reject that runs before anything touches disk, same shape as your snippet.

One deliberate departure: I derived it from `CURATED_PM_FILES`/`SPRINT_GLOBS` via `fnmatch` rather than the hand-written `r"^SPRINT_\d+_PLAN\.md$"`. A regex in `pm_daemon.py` would be a *new* copy of the exact pattern this module exists to hold once — it would have fixed the perf bug by re-committing the bug the PR is about.

Mid-session pickup is preserved: the fast path only rejects, never accepts, so the confirmation against `canonical_pm_files()` still runs for anything that could match. The test that matters is `test_never_rejects_what_canonical_pm_files_accepts` — a false negative here would silently stop syncing a real plan, which is a worse failure than the cost it removes.

_Replied by Claude on Adam's behalf_

File: .claude/hooks/sessionstart-pm-briefing.sh [hunk 2]
```diff
@@ -57,24 +57,54 @@ fi
...
+        "specs/design-plan.md",
+        "specs/SCRUM_BOOTSTRAP_AND_BOARD_PLAN.md",
+        "specs/ATLASSIAN_PM_LINK.md",
+    ]
+
+    def canonical_pm_files(root="."):
+        base = Path(root)
+        found = [p for p in _FALLBACK if (base / p).is_file()]
+        found += sorted(
+            m.relative_to(base).as_posix() for m in base.glob("specs/SPRINT_*_PLAN.md") if m.is_file()
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply .claude/hooks/sessionstart-pm-briefing.sh]: Fixed rather than documented, in `d2f95ab` — took option 1 (inline the numeric key), not option 2.

The reasoning against the comment-only route: a *silently reordered briefing* is exactly the class of difference nobody thinks to check. Documenting it would have made the comment true and left the behaviour divergent, in the one path where there is no second signal that anything is off.

Deep-tail is right about reachability, but the cost of closing it was ~6 lines, and the module already has `test_sprint_plans_sort_numerically` asserting numeric order — so the fallback disagreeing with it was a documented contract with one unenforced implementation.

Pinned by `test_fallback_sorts_sprints_numerically_like_the_module`, which builds a tree with **no** `scripts/pm` to force the fail-open branch, then asserts SPRINT_1 · SPRINT_2 · SPRINT_10.

Follow-on you should know about: chasing this exposed a second divergence in the same fallback that neither of us named — it omitted `SPRINT_0_PLAN.md` from its curated list, so the glob appended it *after* `ATLASSIAN_PM_LINK.md` while the module emits it *before*. Same files, different order. Fixed in `799e315` with a test that compares the whole ordered list instead of just the sprint block.

_Replied by Claude on Adam's behalf_

**Comment by adamtasteslikegood** [comment]
Two of the three fixed in `799e315`; one rebutted. Verified each against the code before touching anything.

### 1. Fallback ordering — confirmed, and it was worse than "only reachable when `scripts/pm` is absent"

Confirmed empirically rather than by reading. Same tree, both paths:

```
MODULE                                    FALLBACK (before fix)
  specs/plan.md                             specs/plan.md
  specs/roadmap.md                          specs/roadmap.md
  specs/planning_notes.md                   specs/planning_notes.md
  specs/design-plan.md                      specs/design-plan.md
  specs/SCRUM_BOOTSTRAP_AND_BOARD_PLAN.md   specs/SCRUM_BOOTSTRAP_AND_BOARD_PLAN.md
  specs/SPRINT_0_PLAN.md   <-- position 6   specs/ATLASSIAN_PM_LINK.md
  specs/ATLASSIAN_PM_LINK.md                specs/SPRINT_0_PLAN.md   <-- position 7
```

Cause: `_FALLBACK` omitted `SPRINT_0_PLAN.md`, so the glob appended it after the curated block instead of the module emitting it inside one. Fixed by mirroring `CURATED_PM_FILES` exactly, plus a dedupe so the doubly-matched entry does not brief twice.

**The second half of your finding is the more useful half.** You are right that `test_hook_and_module_resolve_the_same_set` does not exercise the fallback — it builds a repo that *has* `scripts/pm`, so it only ever tested the import branch. The two fallback tests check membership and sprint sort but never full order, and this divergence sat exactly in the gap between them. `test_fallback_order_matches_the_module_exactly` now compares the whole ordered list against `canonical_pm_files()` on a tree with no `scripts/pm`. Mutation-checked: reverting `_FALLBACK` fails it on precisely the SPRINT_0 position.

One scope limit I wrote into the test docstring rather than leaving implied — it does **not** cover the fallback's dedupe. I checked by mutation: deleting the dedupe guard keeps the test green, because the hook's consumer loop carries a second `seen` set that drops duplicates before output. The dedupe stays for parity with the module's contract, but nothing pins it, and claiming otherwise would have been the same weak-assertion failure I'm fixing on #3316.

### 2. Redundant `filepath.exists()` — keeping it, and it is load-bearing for the denominator

Correct that it is redundant against `canonical_pm_files()`, which already filters on `is_file()`. Keeping it as the race guard you suspected: `considered` is computed once at line 297 and the loop then does network I/O per file, so a file deleted mid-sync would otherwise hit `read_text` inside `sync_to_confluence`.

The part that makes it more than defensive: when it fires, that file lands in neither `synced` nor `failed`, so `len(synced) + len(failed) < len(considered)` and the "Considered N" line reports the gap. That is the exact signal this PR added the denominator for. Removing the guard would convert a visible shortfall into a logged read error.

### 3. `_page_title_for` alias — removed

Agreed, no reason for it. It existed only to avoid touching its single call site. Dropped; `sync_to_confluence` calls `page_title_for` directly. Verified the daemon still imports under its venv.

80/80 in `python3 -m unittest discover -s scripts/pm` (4 skipped, unrelated).

_Replied by Claude on Adam's behalf_

**Comment by adamtasteslikegood** [comment]
Re: the one acceptance criterion Code Reviewer could not determine — *"sync_pm_documents must name how many canonical files it considered, so an omission is visible rather than indistinguishable from success."*

It is met. `scripts/pm/pm_daemon.py`:

```python
considered = canonical_pm_files(workspace_dir)
...
return (
    ". ".join(result)
    + f". Considered {len(considered)} canonical PM file(s). "
    "The PM Daemon has finished updating the documents in Atlassian."
)
```

The denominator is unconditional — it prints on every non-empty result, not only on failure. That is the whole point: the original bug was that `sync_pm_documents` reported only what it *synced*, so four sprint plans that were never on the list read identically to four that synced fine, and the tool returned "Sync successful" for four sprints while ~91 KB of planning never reached Confluence.

Worth naming why this one is hard to verify automatically: it is an assertion about what a *message* makes visible to a human reader, and there is no test that can check "an operator would notice." The mechanical part — that `considered` is the full canonical set — is covered by the suite; the legibility part is a judgment call, and it is Adam's to accept or reject.

_Replied by Claude on Adam's behalf_
